src/ServiceControl.Monitoring/QueueLength/QueueLengthDataStore.cs

```diff
@@ -11,22 +13,57 @@ public QueueLengthDataStore(IQueueLengthCalculator calculator)
             this.calculator = calculator;
         }
 
-        public IEnumerable<KeyValuePair<string, JObject>> Current
+        public Dictionary<string, Dictionary<DateTime, double>> GetQueueLengths(DateTime now)
         {
-            get { return calculator.GetQueueLengths().Select(kvp => new KeyValuePair<string, JObject>(kvp.Key, new JObject(new JProperty("Count", kvp.Value)))); }
+            var results = new Dictionary<string, Dictionary<DateTime, double>>();
+
+            foreach (var endpointName in queueLengths.Keys)
```

Comment:
Iterating over `.Keys` could be replaced by iteration over the concurrent dictionary which does not take locks.